### PR TITLE
Stop publishing tree just to get folder names for the language service

### PR DESCRIPTION
**Pull request by davkean**
We used to publish and look up the project tree to find the effective path of a `<Compile/>` to pass to Roslyn. Originally I did this to handle that fact that some trees _might_ result in different view of the files (think Venus Web Sites where all files end up in App_Code). This ended up being quite complex, slow and due active configurations changing, or multi-targeting, could result in incorrect data being passed to Roslyn. For example, in multi-targeting projects, if a file was conditioned out of the current configuration the folder names for that file would never be found because it would never show be shown in the "Active" tree.

Instead we just get the folder names from the item itself, looking at Link metadata if specified, here's some examples:

```
<Compile Include="File.cs" />                         --> {}
<Compile Include="Foo\File.cs" />                     --> {Foo}
<Compile Include="Foo\Bar\File.cs" />                 --> {Foo, Bar}
<Compile Include="..\Foo\Bar.cs" Link="Bar.cs" />     --> {}
<Compile Include="..\Foo\Bar.cs" Link="Foo\Bar.cs" /> --> {Foo}

```

Fixes: https://github.com/dotnet/roslyn-project-system/issues/1760, https://github.com/dotnet/roslyn-project-system/issues/1655

Tests are coming.

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/project-system @mavasani

**Comment by davkean** [comment]
Side note, it's somewhat frustrating that we need to pass this info to Roslyn instead of it just getting it from the project system, hint, hint @CyrusNajmabadi... :)

**Comment by CyrusNajmabadi** [comment]
sorry, not sure why i was pinged :)

Is there something you'd like me to do?

**Comment by davkean** [comment]
Why is a requirement that we have to pass a bunch of strings to be stored that are only used for specific code actions?

**Comment by CyrusNajmabadi** [comment]
> that are only used for specific code actions?

I wouldn't say that they're *only* for specific code actions :)

I mean, it's just part of the Workspace model that someone can get a Document and ask it for Documenty questions (like "what's your name?" and "what're your folders?").  

That said, TextDocument.Folders is just defined as:

```c#
        public IReadOnlyList<string> Folders => State.Folders;
```

So it looks like it could be something that is computed on demand.  I'd have no problem if such a change was made.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/FileItemServices.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Immutable;
+using System.IO;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    internal static class FileItemServices
+    {
+        public static readonly char[] PathSeparatorCharacters = { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar };
+
+        public static string GetLinkFilePath(IImmutableDictionary<string, string> metadata)
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/FileItemServices.cs]: This seems like its worth a unit test or two.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/FileItemServices.cs]: Yeah tests are coming...

**Pull request merged**